lib/Sema/ConstraintSystem.cpp

```diff
@@ -3099,6 +3031,112 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
   return diagnosed;
 }
 
+bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+    SmallVectorImpl<Solution> &solutions) {
+  if (solutions.empty())
+    return false;
+
+  SolutionDiff solutionDiff(solutions);
+
+  if (diagnoseConflictingGenericArguments(*this, solutionDiff, solutions))
+    return true;
+
+  if (auto bestScore = solverState->BestScore) {
+    solutions.erase(llvm::remove_if(solutions,
+                                    [&](const Solution &solution) {
+                                      return solution.getFixedScore() >
+                                             *bestScore;
+                                    }),
+                    solutions.end());
+
+    if (llvm::all_of(solutions, [&](const Solution &solution) {
+          auto score = solution.getFixedScore();
+          return score.Data[SK_Fix] == 0 && solution.Fixes.empty();
+        }))
+      return false;
+  }
+
+  if (solutions.size() < 2)
+    return false;
+
+  if (diagnoseAmbiguityWithEphemeralPointers(*this, solutions))
+    return true;
+
+
+  // Algorithm is as follows:
+  //
+  // a. Aggregate all of the available fixes based on callee locator;
+  // b. For each ambiguous overload match aggregated fixes and diagnose;
+  // c. Discard all of the fixes which have been already considered
+  //    as part of overload diagnostics;
+  // d. Diagnose remaining (uniqued based on kind + locator) fixes
+  //    iff they appear in all of the solutions.
+
+  using Fix = std::pair<const Solution *, const ConstraintFix *>;
+
+  llvm::SmallSetVector<Fix, 4> fixes;
+  for (auto &solution : solutions) {
+    for (auto *fix : solution.Fixes)
+      fixes.insert({&solution, fix});
+  }
+
+  llvm::MapVector<ConstraintLocator *, SmallVector<Fix, 4>> fixesByCallee;
+
+  for (const auto &entry : fixes) {
+    const auto &solution = *entry.first;
+    const auto *fix = entry.second;
+    auto *calleeLocator = solution.getCalleeLocator(fix->getLocator());
+
+    fixesByCallee[calleeLocator].push_back({&solution, fix});
```

Comment:
Just curious, but at this point is possible for `getCalleeLocator` to return a `nullptr` if it does what happens to the fix and solution?

Reply:
`getCalleeLocator` should never return `nullptr`, its default behaviour (if no other case applies) is to return a locator with the same anchor as the input locator.